File: WQChartsDemo/Objc/View/ButtonsCell.m
// https://github.com/CoderWQYao/WQCharts-iOS
//
// ButtonsCell.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "ButtonsCell.h"

@interface ButtonsCell()

@property (nonatomic, strong) UILabel* titleLabel;
@property (nonatomic, strong) UIView* buttonsWrap;
@property (nonatomic) CGFloat buttonHeight;
@property (nonatomic) CGFloat buttonSpacing;

@end


@implementation ButtonsCell


#pragma mark - Getter

- (UILabel *)titleLabel {
    if(!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont systemFontOfSize:10];
        _titleLabel.textColor = Color_White;
    }
    return _titleLabel;
}

- (UIView *)buttonsWrap {
    if(!_buttonsWrap) {
        _buttonsWrap = UIView.new;
    }
    return _buttonsWrap;
}

- (NSString *)title {
    return self.titleLabel.text;
}

#pragma mark - Setter

- (void)setTitle:(NSString *)title {
    self.titleLabel.text = title;
    [self setNeedsLayout];
}

- (void)setOptions:(NSArray<NSString*>*)options {
    _options = options;
    [self rebuildButtons];
}

- (void)setStyle:(ButtonsCellStyle)style {
    _style = style;
    [self rebuildButtons];
}

#pragma mark - Public

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.buttonHeight = 25;
        self.buttonSpacing = 5;
        self.backgroundColor = UIColor.clearColor;
        
        [self addSubview:self.titleLabel];
        [self addSubview:self.buttonsWrap];
    }
    return self;
}


- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat width = self.bounds.size.width;
    CGFloat height = self.bounds.size.height;
    CGFloat offsetY = 10;
    
    if (self.title.length > 0) {
        UILabel* titleLabel = self.titleLabel;
        [titleLabel sizeToFit];
        titleLabel.frame = CGRectMake(10, offsetY, width - 20, CGRectGetHeight(titleLabel.bounds));
        offsetY = CGRectGetMaxY(self.titleLabel.frame) + 5;
    }
    
    UIView* buttonsWrap = self.buttonsWrap;
    buttonsWrap.frame = CGRectMake(10, offsetY, width - 20, height - offsetY - 10);
    [self layoutButtons];
}

- (CGSize)sizeThatFits:(CGSize)size {
    CGFloat height = 10;
    
    if (self.title.length > 0) {
        UILabel* titleLabel = self.titleLabel;
        [titleLabel sizeToFit];
        height += titleLabel.bounds.size.height;
    }
    
    NSInteger buttonCount = self.buttonsWrap.subviews.count;
    if (buttonCount > 0) {
        NSInteger buttonRowCount = buttonCount / 2 + (buttonCount % 2 > 0 ? 1 : 0);
        height += buttonRowCount * (self.buttonHeight + self.buttonSpacing);
    }
    
    height += 10;
    return CGSizeMake(size.width, MIN(height, size.height));
}

#pragma mark - Internal

- (void)rebuildButtons {
    
    UIView* buttonsWrap = self.buttonsWrap;
    for (UIView* button in buttonsWrap.subviews) {
        [button removeFromSuperview];
    }
    
    NSArray<NSString*>* options = self.options;
    
    if(options==nil) {
        return;
    }
    
    NSInteger count = options.count;
    for (NSInteger i=0; i<count; i++) {
        NSString* option = options[i];
        UIButton* button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.tag = i;
        button.titleLabel.font = [UIFont systemFontOfSize:9];
        [button setTitle:option forState:UIControlStateNormal];
        [self adjustButtonStyle:button isSelected:NO];
        [buttonsWrap addSubview:button];
    }
    [self setNeedsLayout];
}

- (void)layoutButtons {
    UIView* buttonsWrap = self.buttonsWrap;
    NSInteger buttonCount = buttonsWrap.subviews.count;
    if(buttonCount==0) {
        return;
    }
    
    CGFloat width = buttonsWrap.bounds.size.width;
    CGFloat buttonWidth = buttonCount > 1 ? (width - 5) / 2 : width;
    CGFloat buttonHeight = self.buttonHeight;
    CGFloat buttonSpacing = self.buttonSpacing;
    for (NSInteger i=0; i<buttonCount; i++) {
        UIView* button = buttonsWrap.subviews[i];
        NSInteger column = i % 2;
        NSInteger row = i / 2;
        button.frame = CGRectMake(column * (buttonWidth + buttonSpacing),row * (buttonHeight + buttonSpacing), buttonWidth, buttonHeight);
        button.layer.cornerRadius = buttonHeight / 5;
    }
}

- (void)adjustButtonStyle:(UIButton*)button isSelected:(BOOL)isSelected {
    button.selected = isSelected;
    
    if (self.style == ButtonsCellStyleRadio || self.style == ButtonsCellStyleCheckBox) {
        if (isSelected) {
            button.backgroundColor = Color_White;
            [button setTitleColor:Color_Tint forState:UIControlStateNormal];
            button.layer.borderWidth = 1;
            button.layer.borderColor = Color_White.CGColor;
        } else {
            button.backgroundColor = nil;
            [button setTitleColor:Color_White forState:UIControlStateNormal];
            button.layer.borderWidth = 1;
            button.layer.borderColor = Color_White.CGColor;
        }
    } else {
        button.backgroundColor = Color_Orange;
        [button setTitleColor:Color_White forState:UIControlStateNormal];
        button.layer.borderWidth = 0;
        button.layer.borderColor = nil;
    }
    
}

@end

File: WQChartsDemo/Swift/View/ButtonsCell.swift
// https://github.com/CoderWQYao/WQCharts-iOS
//
// ButtonsCell.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class ButtonsCell: UIView {
    
    enum Style {
        case Normal,Radio,Checkbox
    }
    
    // MARK: - Property
    
    open lazy var titleLabel: UILabel = {
        let titleLabel = UILabel()
        titleLabel.font = UIFont.systemFont(ofSize: 10)
        titleLabel.textColor = Color_White
        return titleLabel
    }()
    
    open lazy var buttonsWrap: UIView = {
        let buttonsWrap = UIView()
        return buttonsWrap
    }()
    
    open var title: String? {
        get {
            titleLabel.text
        }
        set {
            titleLabel.text = newValue
            setNeedsLayout()
        }
    }
    
    open var options: [String]? {
        didSet {
            rebuildButtons()
        }
    }
    
    open var style: Style = .Normal {
        didSet {
            rebuildButtons()
        }
    }
    
    private var buttonHeight: CGFloat = 25
    private var buttonSpacing: CGFloat = 5
    
    // MARK: - Public
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        backgroundColor = .clear
        
        addSubview(titleLabel)
        addSubview(buttonsWrap)
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let width = self.bounds.width
        let height = self.bounds.height
        var offsetY: CGFloat = 10
        
        let titleLabel = self.titleLabel
        titleLabel.sizeToFit()
        titleLabel.frame = CGRect(x: CGFloat(10), y: offsetY, width: width - 20, height: titleLabel.bounds.height)
        offsetY = titleLabel.frame.maxY + 5
        
        let buttonsWrap = self.buttonsWrap
        buttonsWrap.frame = CGRect(x: 10, y: offsetY, width: width - 20, height: height - offsetY - 10)
        layoutButtons()
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        var height: CGFloat = 10
        
        let titleLabel = self.titleLabel
        titleLabel.sizeToFit()
        height += titleLabel.bounds.height
        
        let buttonCount = buttonsWrap.subviews.count
        if buttonCount > 0 {
            let buttonRowCount = buttonCount / 2 + (buttonCount % 2 > 0 ? 1 : 0)
            height += CGFloat(buttonRowCount) * (buttonHeight + buttonSpacing)
        }
        
        height += 10
        return CGSize(width: size.width, height: min(height, size.height))
    }
    
    // MARK: - Internal
    
    func rebuildButtons() {
        for button in buttonsWrap.subviews {
            button.removeFromSuperview()
        }
        
        guard let options = options else {
            return
        }
        
        let count: Int = options.count
        for i in 0..<count {
            let option = options[i]
            let button = UIButton(type: .custom)
            button.tag = i
            button.titleLabel?.font = UIFont.systemFont(ofSize: 9)
            button.setTitleColor(Color_White, for: .normal)
            button.setTitle(option, for: .normal)
            adjustButtonStyle(button, false)
            buttonsWrap.addSubview(button)
        }
        
        setNeedsLayout()
    }
    
    func layoutButtons() {
        let buttonsWrap = self.buttonsWrap
        let buttonCount = buttonsWrap.subviews.count
        if buttonCount==0 {
            return
        }
        
        let width = buttonsWrap.bounds.width
        let buttonWidth: CGFloat = buttonCount > 1 ? (width - 5) / 2 : width
        let buttonHeight = self.buttonHeight
        let buttonSpacing = self.buttonSpacing
        for i in 0..<buttonCount {
            let button = buttonsWrap.subviews[i]
            let column = i % 2
            let row = i / 2
            button.frame = CGRect(x: CGFloat(column) * (buttonWidth + buttonSpacing), y: CGFloat(row) * (buttonHeight + buttonSpacing), width: buttonWidth, height: buttonHeight)
            button.layer.cornerRadius = buttonHeight / 5;
        }
    }
    
    func adjustButtonStyle(_ button: UIButton, _ isSelected: Bool) {
        button.isSelected = isSelected
        
        if self.style == .Radio || self.style == .Checkbox {
            if isSelected {
                button.backgroundColor = Color_White
                button.setTitleColor(Color_Tint, for: .normal)
                button.layer.borderWidth = 1;
                button.layer.borderColor = Color_White.cgColor
            } else {
                button.backgroundColor = nil
                button.setTitleColor(Color_White, for: .normal)
                button.layer.borderWidth = 1;
                button.layer.borderColor = Color_White.cgColor
            }
        } else {
            button.backgroundColor = Color_Orange
            button.setTitleColor(Color_White, for: .normal)
            button.layer.borderWidth = 0
            button.layer.borderColor = nil
        }
        
    }
    
}
